staker/challenge-cache/cache.go

```diff
@@ -0,0 +1,242 @@
+	return &Cache{
+		baseDir: baseDir,
+	}, nil
+}
+
+// Key for cache lookups includes the wavm module root of a challenge, as well
+// as the heights for messages and big steps as needed.
+type Key struct {
+	RollupBlockHash common.Hash
+	WavmModuleRoot  common.Hash
+	MessageHeight   uint64
+	StepHeights     []uint64
+}
+
+// Get a list of hashes from the cache from index 0 up to a certain index. Hashes are saved as files in the directory
+// hierarchy for the cache. If a file is not present, ErrNotFoundInCache
+// is returned.
+func (c *Cache) Get(
+	lookup *Key,
+	numToRead uint64,
+) ([]common.Hash, error) {
+	fName, err := determineFilePath(c.baseDir, lookup)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := os.Stat(fName); err != nil {
+		log.Warn("Cache miss", "fileName", fName)
+		return nil, ErrNotFoundInCache
+	}
+	log.Debug("Cache hit", "fileName", fName)
+	f, err := os.Open(fName)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Error("Could not close file after reading", "err", err, "file", fName)
+		}
+	}()
+	return readHashes(f, numToRead)
+}
+
+// Put a list of hashes into the cache.
+// Hashes are saved as files in a directory hierarchy for the cache.
+// This function first creates a temporary file, writes the hashes to it, and then renames the file
+// to the final directory to ensure atomic writes.
+func (c *Cache) Put(lookup *Key, hashes []common.Hash) error {
+	// We should error if trying to put 0 hashes to disk.
+	if len(hashes) == 0 {
+		return ErrNoHashes
+	}
+	fName, err := determineFilePath(c.baseDir, lookup)
+	if err != nil {
+		return err
+	}
+	// We create a tmp file to write our hashes to first. If writing fails,
+	// we don't want to leave a half-written file in our cache directory.
+	// Once writing succeeds, we rename in an atomic operation to the correct file name
+	// in the cache directory hierarchy.
+	tmp, err := os.MkdirTemp(c.baseDir, "tmpdir")
+	if err != nil {
+		return err
+	}
+	tmpFName := filepath.Join(tmp, fName)
+	dir := filepath.Dir(tmpFName)
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return fmt.Errorf("could not make tmp directory %s: %w", dir, err)
+	}
+	f, err := os.Create(tmpFName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Error("Could not close file after writing", "err", err, "file", fName)
+		}
+	}()
+	if err := writeHashes(f, hashes); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Dir(fName), os.ModePerm); err != nil {
+		return fmt.Errorf("could not make file directory %s: %w", fName, err)
+	}
+	// If the file writing was successful, we rename the file from the tmp directory
+	// into our cache directory. This is an atomic operation.
+	// For more information on this atomic write pattern, see:
+	// https://stackoverflow.com/questions/2333872/how-to-make-file-creation-an-atomic-operation
+	return os.Rename(tmpFName /*old */, fName /* new */)
+}
```

Comment:
can be a separate PR - add a Prune function that will get as a parameter finalized message number and will prune all dirs for earlier message